build/assets/scripts/high-performance-hooks.sh

```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+pid=$(jq '.pid' /dev/stdin)
+mask=$1
+ns=$(ip netns identify "${pid}")
+
+[[ -n "${pid}" ]] && [[ -n "${mask}" ]] && [[ -n "${ns}" ]] || exit 0
+
+ip netns exec "${ns}" mount -o remount,rw /sys || exit 0
+ip netns exec "${ns}" find /sys/devices -type f -name rps_cpus -exec sh -c "echo ${mask} | cat > {}" \;
```

Comment:
After the @fromanirh comment regarding error handling, IMHO we should add -e and add a trap that will call to `ip netns exec "${ns}" mount -o remount,ro /sys`, because if the command
`ip netns exec "${ns}" find /sys/devices -type f -name rps_cpus -exec sh -c "echo ${mask} | cat > {}" \;` will fail but the remount command after it will succeed, the hook will exit with 0 code, and from the cri-o perspective it will mean that hook succeeded, so a user can get unexpected results

Reply:
This is on purpose, I am interested to return as less errors as possible, if the hook fails I want the pod to continue to run.  I prefer the pod to be running with the rps mask not set, than not running at all

Reply:
hm if you expect some functionality and you do not get it silently it can lead to unexpected behavior, but I will not block on it

Reply:
yeah, is a hack. we will push for a sane RPS interface, sadly is the best it can be done at current time.

Reply:
The question is: even if we properly exit with error and we log the reason, where do the log ends and how we can check it? If it ends up somewhere we can check (crio logs?) it's worth improving. Even something trivial like well defined exit codes for the failure conditions can be enough.

Reply:
In general, what happens if a hook exits with error?

Reply:
The point is we don't actually want to error out. We can use the Logger to send the logs to the syslog, but is not necessary since is the caller's responsibility to catch the stderr of the script and log it.

The current approach does the best it can:
1. if it can't remount the sysfs as rw, exits with 0 so it will let the pod continue to run, it makes no sense to punish some random pod because of this.
2. At this point we go for the rps_cpus files write, even if it doesn't succeed, we still need to try to unmount and the pod should still continue to run.
3. The last try to remount as ro is tricky, we shouldn't let an unprivileged pod to run with sysfs rw. But if so, since is the last script line, it will return a non-zero error code and the pod will not run.

All of the above is (should be) logged by the caller.

Reply:
> The point is we don't actually want to error out. We can use the Logger to send the logs to the syslog, but is not necessary since is the caller's responsibility to catch the stderr of the script and log it.

True. Do we know where do these logs end? crio logs?

> The current approach does the best it can:
> 
>     1. if it can't remount the sysfs as rw, exits with 0 so it will let the pod continue to run, it makes no sense to punish some random pod because of this.

Well, this is debatable, and that's indeed why we are discussing. Guaranteed pods which opt in to the behaviour we are implementing here expect some preconditions. If we fail to set up the RPS mask these preconditions aren't guaranteed anymore. At this point keep going with the pod start up may or may be not the right thing to do but what @cynepco3hahue AFAIU raised (and if so I totally agree) is that we should never *silently* fail to comply with the pod expectations.

So if the hook cannot properly set up the environment the guarantees aren't met anymore and this fact should be evident somewhere. Which is indeed responsability of the caller (crio) to log, but we need to know where this happens and how we can check later on.

Reply:
But this hook is called for all the pods in the system, right? I see two issues:

1) Last time we used cri-o code approach instead of wrapper, because the OCP maintainers were worried about scalability (what happens when huge amount of pods starts at the same time). This we can't probably fix atm.
2) A failing hook prevents the start of pods, we must be super careful to not compromise the whole node

Reply:
> In general, what happens if a hook exits with error?

The container ends up in status: CreateContainerError
And in the pod status we see the exact error:
  Warning  Failed          1s               kubelet, ocp47-worker-1.demo.lab  Error: container create failed: time="2020-11-26T08:26:48Z" level=error msg="container_linux.go:366: starting container process caused: process_linux.go:472: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: /usr/local/bin/high-performance-hooks.sh: line 13: /bin/d: Read-only file system\n"

Reply:
> True. Do we know where do these logs end? crio logs?
> 

The pod event status has the detailed output, please see the above exampe

Reply:
Thanks, container ending up in `CreateContainerError` matches my expectations and it's indeed better than silently fail to meet the expectations.

Reply:
> Well, this is debatable, and that's indeed why we are discussing. Guaranteed pods which opt in to the behaviour we are implementing here expect some preconditions. If we fail to set up the RPS mask these preconditions aren't guaranteed anymore. At this point keep going with the pod start up may or may be not the right thing to do but what @cynepco3hahue AFAIU raised (and if so I totally agree) is that we should never _silently_ fail to comply with the pod expectations.
> 

We can opt to return error in this case, that means is possible no container would start and the node will appear as "Not Ready". But if the consensus is that is a critical matter I will change it to return error.


> So if the hook cannot properly set up the environment the guarantees aren't met anymore and this fact should be evident somewhere. Which is indeed responsability of the caller (crio) to log, but we need to know where this happens and how we can check later on.

If the script is returning with exit 0, no interim error is logged and sensing something to syslog does not work either.

Reply:
> But this hook is called for all the pods in the system, right? I see two issues:
> 
> 1. Last time we used cri-o code approach instead of wrapper, because the OCP maintainers were worried about scalability (what happens when huge amount of pods starts at the same time). This we can't probably fix atm.

Nope, we can't, it will be a short delay on each pod startup. The "price" to pay for the low-latency.
But, as you pointed out, the systems requiring RT are quite static, running a few pods.

> 2. A failing hook prevents the start of pods, we must be super careful to not compromise the whole node
Exactly what I said! 

We have 2 points of view and we need a decision:
If the pod hook fails to do its job it will affect the Guaranteed Pods latency.
So what should we do?
 - Deny it to run can and will compromise the node, for example it may not allow an operator pod to run. 
 - Allow it to run anyway will compromise the Guaranteed Pods latency.

Reply:
> Thanks, container ending up in `CreateContainerError` matches my expectations and it's indeed better than silently fail to meet the expectations.

Right, we don't have a security risk.

Our Achilles' heel remains if the writing the rps_cpus fails and the remount to ro succeeded. In this case the hook returns with 0. I found no way to log something if the hook succeeded